cpp/src/arrow/io/file.cc

```diff
@@ -517,6 +517,10 @@ Status MemoryMappedFile::WriteAt(int64_t position, const void* data, int64_t nby
   }
 
   RETURN_NOT_OK(memory_map_->Seek(position));
+  if (nbytes + memory_map_->position() > memory_map_->size()) {
```

Comment:
We should probably do this check before seeking.

Reply:
Absolutely. I added a check for it.